---
title: "Cultures: 16S rRNA gene amplicon sequencing"
author: "L. Bettera & M. Dreier"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_document
---


```{r ngs packages, include=FALSE}
source('genfuncs.R')
Rdeps = c("BiocManager", "openxlsx",
          "plyr", "tidyverse", "vegan",
          "stats", 'ggtext', 'magrittr',
          'patchwork', 'rjson')
Biocdeps = c("phyloseq")
checkpackages(Rdeps)
checkBiocpackages(Biocdeps)
```

This R script will analyze 16S results from ASV counts to graphical visualization
mainly using phyloseq http://joey711.github.io/phyloseq/import-data.html

## Cultures

```{r  cultures NGS data import}
# import raw sequences that were already quality filtered as described in Dreier et al 2022
fp <- file.path(getwd(), "..", "data", "cultures", "ngs", "ASV_table.xlsx")
cult_physeq <- load_ngs_data(fp)
```


```{r}
# Counts to relative abundance
# ASV counts transformed to relative abundance (%)
# Note: raw sequences were already quality filtered as described in Dreier et al 2022
# thus taxa are not further trimmed for ordination and diversity analysis
# since even low abundant species are informative
cult_physeqR  = transform_sample_counts(cult_physeq, function(x) x/sum(x)*100 )
```


```{r}
# font sizes for plots
yaxis_tick_label_size = 10
yaxis_label_size = 10

title_size = 14
subtitle_size = 14
xaxis_tick_label_size = 12
```


### Diversity
```{r cultures diversity}
# Summarize diversity values in table
metadata <- as(sample_data(cult_physeq), "data.frame")
richness_table <- estimate_richness(cult_physeq)
richness_table["Sample"] <- metadata["sample_subtype"]
richness_table["replicate"] <- metadata["replicate"]
# Reorder columns
richness_table <- richness_table[, c(10, 11, 1, 2, 3, 4, 5, 6, 7, 8 , 9)]
richness_table
# write to excel
write.xlsx(richness_table, '../tables/cultures_richness_table.xlsx', colNames = TRUE, rowNames = TRUE, asTable = FALSE, keepNA = FALSE, na.string = "NaN")

# Beta diversity Permanova for all samples
adonis2(distance(cult_physeq, method="bray") ~ sample_subtype, data = metadata, permutations = 9999)

# filter for cultures
cultures_only_phy <- subset_samples(cult_physeq, sample_type == "eRWC")
# cultures metadata
c_metadata <- as(sample_data(cultures_only_phy), "data.frame")
# Beta diversity Permanova
adonis2(distance(cultures_only_phy, method="bray") ~ sample_subtype, data = c_metadata, permutations = 9999)
```

### Relative abundances

```{r cultures data preparation for plotting}
# percentage thresholds used for filtering
abund_thresh <- 1.0
sp_res_th <- 0.6

# Taxa agglomeration----
# Differently to Ord. & Div. plots, in bar charts a taxa filter is necessary to show the most abundant species
# Let's agglomerate at the Species level using the tax_glom function from phyloseq
# It requires "Kingdom" to be the first column of our tax_table
# (see probl. solved https://github.com/joey711/phyloseq/issues/1551)
tax_table(cult_physeqR) <- tax_table(cult_physeqR)[,c(2:10,1)] # Re-order the columns

# Let's agglomerate at the Species level
# i.e. different ASV belonging to the same species will be agglomerated
# their relative abundance will be summed
# further the data frame is melted to long format for filtering and plotting
cult_physeqR_tax <- cult_physeqR %>%
  tax_glom(taxrank = "Species") %>%
  psmelt()

# Select only the natural whey and the enriched raw milk-whey culture samples 
cult_physeqR_tax_samples = subset(cult_physeqR_tax, sample_type %in% c("eRWC", "NWC"))

# create a mapper to replace old lactobacilli names in table
fp = file.path(getwd(), "..", "data", "species_dict.json")
species_dict <- fromJSON(file = fp, simplify = TRUE)

# In order to get groups in the plots add a group category and rename replicates
cult_physeqR_tax_samples <- cult_physeqR_tax_samples %>%
  mutate(sample_subtype = substr(sample_ID, 1, nchar(sample_ID) - 2))  %>%
  mutate(replicate = replicate - 1)

# For plotting it looks nicer when the species names are separated by spaces
cult_physeqR_tax_samples <- cult_physeqR_tax_samples %>%
    mutate(Species = str_replace(Species, "_", " ")) %>%
    # update new lactobacilli names
    mutate(Species = recode_factor(Species, !!!species_dict)) %>%
    # change Species to sp. for unknown species names
    mutate(Species = str_replace(Species, "Species", "sp."))

# group data frame by Species, calculate mean rel. abundance
means <- ddply(cult_physeqR_tax_samples, ~Species, function(x) c(mean = mean(x$Abundance)))

# get the species names with high and low average relative abundance, respectively
# Species whose rel. abundance is higher than 1% (or threshold [abund_thresh] defined above)
high_ab_sp <- means[means$mean >= abund_thresh, "Species"]
# Species whose rel. abundance is less than 1%
low_ab_sp <- means[means$mean < abund_thresh, "Species"]

# Create two data frames with high and low avg. abundances
cult_physeqR_tax_high <- cult_physeqR_tax_samples[cult_physeqR_tax_samples$Species %in% high_ab_sp,]
cult_physeqR_tax_low <- cult_physeqR_tax_samples[cult_physeqR_tax_samples$Species %in% low_ab_sp,]

low_abund <- cult_physeqR_tax_low
lowabund_str <- paste(c("<", abund_thresh, "% avg. abundance"), collapse = " ")
low_abund$Species <- lowabund_str
cult_physeqR_tax_high_full <- rbind(cult_physeqR_tax_high, low_abund)


# convert Species to a character vector from a factor because R
cult_physeqR_tax_low$Species <- as.character(cult_physeqR_tax_low$Species)

# group data frame by Species, calculate max rel. abundance
maxs <- ddply(cult_physeqR_tax_low, ~Species, function(x) c(max = max(x$Abundance)))

# find Species whose max. abundance is less than 0.6% [threshold "sp_res_th" defined above]
remainder <- maxs[maxs$max <= sp_res_th,]$Species

# change their name to "Other species"
cult_physeqR_tax_low[cult_physeqR_tax_low$Species %in% remainder,]$Species <- 'Other species'
```


```{r}
# create nicely formatted guides with the species names
ordered_species_list <- ordered_species_guides(cult_physeqR_tax_low$Species, "Other species")
ordered_species <- ordered_species_list[[1]]
ordered_sp_labels <- ordered_species_list[[2]]

ha_ordered_species_list <- ordered_species_guides(cult_physeqR_tax_high$Species, lowabund_str)
ha_ordered_species <- ha_ordered_species_list[[1]]
ha_ordered_sp_labels <- ha_ordered_species_list[[2]]

# sample_ID levels
sampleid_vector <- c("NWC.2", "NWC.3", "NWC.4",
                     "eRWC.y.2", "eRWC.y.3", "eRWC.y.4",
                     "eRWC.o.2", "eRWC.o.3", "eRWC.o.4",
                     "eRWC.H.y.2", "eRWC.H.y.3", "eRWC.H.y.4",
                     "eRWC.H.o.2", "eRWC.H.o.3", "eRWC.H.o.4",
                     "eRWC.HS.y.2", "eRWC.HS.y.3", "eRWC.HS.y.4",
                     "eRWC.HS.o.2", "eRWC.HS.o.3", "eRWC.HS.o.4")

# Colors are set manually
species_colors <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", 
                    "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

# Define font size for both plots here
yaxis_tick_label_size = 10
yaxis_label_size = 10
xaxis_tick_label_size = 12
Xaxis_rot = 0
df_sorter = c("Species", "sample_subtype")
x_labels = c(".1",".2",".3")

# High abundant species plot
p1 <- draw_ha_stacked_plot(df = cult_physeqR_tax_high_full,
                           sampleid_vector = sampleid_vector,
                           x_factor = "sample_ID",
                           x_labels = x_labels,
                           species_colors = species_colors, 
                           df_sorter = df_sorter,
                           wrap_group = c("~sample_subtype"),
                           ordered_sp_labels = ha_ordered_sp_labels,
                           abund_thresh = abund_thresh,
                           sp_res_th = sp_res_th,
                           ordered_species = ha_ordered_species
                           )

# Low abundant species plot
p2 <- draw_la_stacked_plot(df = cult_physeqR_tax_low,
                           sampleid_vector = sampleid_vector,
                           x_labels = x_labels,
                           species_colors = species_colors,
                           ordered_sp_labels = ordered_sp_labels,
                           x_factor = "sample_ID",
                           ordered_species = ordered_species,
                           wrap_group = c("~sample_subtype"),
                           abund_thresh = abund_thresh,
                           sp_res_th = sp_res_th
                           )
```




```{r create Figure 4, fig.width=12,fig.height=10}
fig4 <- (p1 / p2) + 
              plot_annotation(tag_levels = "A") &
              theme(plot.tag = element_text(size = 18, face = "bold"))
                  
ggsave(filename = file.path(getwd(), "..", "figures", "Figure_4.jpeg"), height = 10 , width = 10 * 1.2)
fig4
```
